### PR TITLE
Improve mismatched delimiter message

### PR DESCRIPTION
**Pull request by viniciusmuller**
This PR uses the `start` and `end` information contained in tokens to show a more precise message on mismatched delimiters error.

I'm opening it as a draft since there's still some tests missing implementations and I'll refactor the implementation a bit before fully opening for review.

@josevalim would you mind taking a look at the test cases and giving some feedback?

# Preview

Screenshot from 2023-09-13 11-27-44

Screenshot from 2023-09-13 11-28-38

Screenshot from 2023-09-13 11-29-59

### CODE REVIEW

File: lib/elixir/test/elixir/kernel/diagnostics_test.exs [hunk 1]
```diff
@@ -10,6 +10,305 @@ defmodule Kernel.DiagnosticsTest do
...
+          #{String.duplicate("\n", 6)}
+            b )
+          """,
+          MismatchedDelimiterError
+        )
+
+      assert output == expected
+    end
+
+    test "trims lines that are too long (> 60 chars)" do
```
Comment (josevalim) [inline lib/elixir/test/elixir/kernel/diagnostics_test.exs:259]: What happens when only one of the delimiters is too long ahead?

File: lib/elixir/test/elixir/kernel/diagnostics_test.exs [hunk 2]
```diff
@@ -10,6 +10,305 @@ defmodule Kernel.DiagnosticsTest do
...
+
+      assert output == expected
+
+      expected = """
+      ** (MismatchedDelimiterError) mismatched delimiter found on nofile:403:5:
+           error: unexpected token: )
+           HINT: the "[" on line 1 is missing terminator "]"
+           │
+         1 │ [ a,
+           │ └ unclosed delimiter
```
Comment (josevalim) [inline lib/elixir/test/elixir/kernel/diagnostics_test.exs:211]: Do you think it is worth showing one line before the opening and one line before the closing? Or would that be just additional noise? It may be helpful for the closing one because often it is just the closing token on its own line.

File: lib/elixir/test/elixir/kernel/diagnostics_test.exs (hunk 1 above)
Comment (josevalim) [reply lib/elixir/test/elixir/kernel/diagnostics_test.exs:259]: And what happens if it is too many lines and too long lines?

File: lib/elixir/lib/exception.ex [hunk 3]
```diff
@@ -949,28 +951,161 @@ defmodule MismatchedDelimiterError do
...
+       )
+       when start_line == end_line do
+    max_digits = digits(end_line)
+    general_padding = max(2, max_digits) + 1
+    padding = n_spaces(general_padding)
+
+    line = Enum.at(lines, end_line - 1)
+    formatted_line = "#{line_padding(end_line, max_digits)}#{end_line} │ #{line}"
+
+    mismatched_closing_line = n_spaces(start_column - 1) <> red("│") <> mismatched_closing_delimiter(abs(end_column - start_column))
```
Comment (josevalim) [inline lib/elixir/lib/exception.ex:1044]: Attempt to rely less on string concatenation and instead build lists which we will interpolate.

File: lib/elixir/lib/exception.ex [hunk 4]
```diff
@@ -949,28 +951,161 @@ defmodule MismatchedDelimiterError do
...
+    do: String.duplicate(" ", end_column - 1) <> red("└ mismatched closing delimiter")
+
+  defp unclosed_delimiter(start_column),
+    do: String.duplicate(" ", start_column - 1) <> red("└ unclosed delimiter")
+
+  defp pad_message(message, padding), do: String.replace(message, "\n", "\n #{padding}")
+
+  defp red(string) do
+    if IO.ANSI.enabled?() do
+      IO.ANSI.red() <> string <> IO.ANSI.reset()
```
Comment (josevalim) [inline lib/elixir/lib/exception.ex:1103]: In the same line:

```suggestion
      [IO.ANSI.red(), string, IO.ANSI.reset()]
```

File: lib/elixir/lib/exception.ex [hunk 5]
```diff
@@ -949,28 +951,161 @@ defmodule MismatchedDelimiterError do
...
+      end
+    end)
+    |> Enum.join("\n")
+  end
+
+  defp mismatched_closing_delimiter(end_column),
+    do: String.duplicate(" ", end_column - 1) <> red("└ mismatched closing delimiter")
+
+  defp unclosed_delimiter(start_column),
+    do: String.duplicate(" ", start_column - 1) <> red("└ unclosed delimiter")
```
Comment (josevalim) [inline lib/elixir/lib/exception.ex:1097]:
```suggestion
    do: [String.duplicate(" ", end_column - 1) | red("└ mismatched closing delimiter")]

  defp unclosed_delimiter(start_column),
    do: [String.duplicate(" ", start_column - 1) | red("└ unclosed delimiter")]
```

File: lib/elixir/lib/exception.ex [hunk 6]
```diff
@@ -949,28 +951,161 @@ defmodule MismatchedDelimiterError do
...
+          """
+          #{line_padding}#{line_number} │ #{line}
+           #{padding}│ #{mismatched_closing_delimiter(end_column)}\
+          """
+
+        true ->
+          "#{line_padding}#{line_number} │ #{line}"
+      end
+    end)
+    |> Enum.join("\n")
```
Comment (josevalim) [inline lib/elixir/lib/exception.ex:1090]:
```suggestion
    |> Enum.intersperse("\n")
```

File: lib/elixir/lib/exception.ex [hunk 7]
```diff
@@ -949,28 +951,161 @@ defmodule MismatchedDelimiterError do
...
+  end
+
+  defp line_range(lines, {start_line, start_column}, {end_line, end_column}, padding, max_digits) do
+    start_line = start_line - 1
+    end_line = end_line - 1
+
+    lines
+    |> Enum.slice(start_line..end_line)
+    |> Enum.zip(start_line..end_line)
+    |> Enum.map(fn {line, line_number} ->
```
Comment (josevalim) [inline lib/elixir/lib/exception.ex:1066]:
```suggestion
    |> Enum.zip_with(start_line..end_line, fn line, line_number ->
```

File: lib/elixir/lib/exception.ex [hunk 8]
```diff
@@ -949,28 +951,161 @@ defmodule MismatchedDelimiterError do
...
+
+        true ->
+          "#{line_padding}#{line_number} │ #{line}"
+      end
+    end)
+    |> Enum.join("\n")
+  end
+
+  defp mismatched_closing_delimiter(end_column),
+    do: String.duplicate(" ", end_column - 1) <> red("└ mismatched closing delimiter")
```
Comment (josevalim) [inline lib/elixir/lib/exception.ex:1094]: Should we use `n_spaces` here and below?

File: lib/elixir/lib/exception.ex [hunk 9]
```diff
@@ -949,28 +951,161 @@ defmodule MismatchedDelimiterError do
...
+    max_digits = digits(end_line)
+    general_padding = max(2, max_digits) + 1
+    padding = n_spaces(general_padding)
+
+    relevant_lines =
+      if end_line - start_line < @max_lines_shown do
+        line_range(lines, start_pos, end_pos, padding, max_digits)
+      else
+        first_line = Enum.at(lines, start_line - 1)
+        last_line = Enum.at(lines, end_line - 1)
```
Comment (josevalim) [inline lib/elixir/lib/exception.ex:987]: It seems we assume those lines indeed exist, right? So it is best to avoid nils propagating:

```suggestion
        first_line = Enum.fetch!(lines, start_line - 1)
        last_line = Enum.fetch!(lines, end_line - 1)
```

**Review by josevalim — COMMENTED** [review]
Looking great, an initial review is in!

File: lib/elixir/test/elixir/kernel/diagnostics_test.exs (hunk 2 above)
Comment (viniciusmuller) [reply lib/elixir/test/elixir/kernel/diagnostics_test.exs:211]: I think that's valid, especially for the closing one, I'll give it a try and see how it looks like

File: lib/elixir/test/elixir/kernel/diagnostics_test.exs (hunk 1 above)
Comment (viniciusmuller) [reply lib/elixir/test/elixir/kernel/diagnostics_test.exs:259]: The idea is to create a helper function that given a line and the column the error is in, trims it up to a point where `column + delimiter_threshold < 80`, where delimiter threshold is a constant of value around 20 (the length of the delimiter message) so that we always trim it showing the most we can while staying under 80 chars and still showing the delimiters for the line, WDYT?

File: lib/elixir/test/elixir/kernel/diagnostics_test.exs (hunk 2 above)
Comment (viniciusmuller) [reply lib/elixir/test/elixir/kernel/diagnostics_test.exs]: It's currently looking like this, wdyt?

### Code

```ex
MyLongModuleWithManyFeatures.call(fn a, b, c ->
  call_svc_1(:a)
  call_svc_2(:b)
  call_svc_3(:c)
  call_svc_4(:d)
  call_svc_5(:e)
)
```

### Error
```
** (MismatchedDelimiterError) mismatched delimiter found on unclosed.ex:7:1:
    error: unexpected token: )
    HINT: the "fn" on line 1 is missing terminator "end"
    │
  1 │ MyLongModuleWithManyFeatures.call(fn a, b, c ->
    │                                   └ unclosed delimiter
    │ ...
  6 │   call_svc_5(:e)
  7 │ )
    │ └ mismatched closing delimiter
    │
    └─ unclosed.ex:7:1
    (elixir 1.16.0-dev) lib/code.ex:1463: Code.require_file/2

```

Comment (josevalim) [reply lib/elixir/test/elixir/kernel/diagnostics_test.exs]: Good point about the line numbers, we can try without adding the extra line and see where it leads us.

File: lib/elixir/test/elixir/kernel/diagnostics_test.exs (hunk 1 above)
Comment (josevalim) [reply lib/elixir/test/elixir/kernel/diagnostics_test.exs:245]: Excellent. Let's make sure we also include the examples from the questions above in the tests. :)

Comment (josevalim) [reply lib/elixir/test/elixir/kernel/diagnostics_test.exs:245]: If you haven't yet!

File: lib/elixir/lib/exception.ex [hunk 10]
```diff
@@ -949,28 +951,174 @@ defmodule MismatchedDelimiterError do
...
+    line = Enum.fetch!(lines, end_line - 1)
+    formatted_line = "#{line_padding(end_line, max_digits)}#{end_line} │ #{line}"
+
+    mismatched_closing_line =
+      [n_spaces(start_column - 1), red("│"), mismatched_closing_delimiter(end_column - start_column)]
+
+    unclosed_delimiter_line =
+      [padding, " │ ", n_spaces(start_column - 1), unclosed_delimiter(start_column)]
+
+    below_line = "#{padding} │ #{mismatched_closing_line}\n#{unclosed_delimiter_line}"
```
Comment (josevalim) [inline lib/elixir/lib/exception.ex:1061]:
```suggestion
    below_line = [padding}, " │ ", mismatched_closing_line, "\n", unclosed_delimiter_line]
```

File: lib/elixir/lib/exception.ex [hunk 11]
```diff
@@ -949,28 +951,174 @@ defmodule MismatchedDelimiterError do
...
+         file,
+         lines
+       )
+       when start_line == end_line do
+    max_digits = digits(end_line)
+    general_padding = max(2, max_digits) + 1
+    padding = n_spaces(general_padding)
+
+    line = Enum.fetch!(lines, end_line - 1)
+    formatted_line = "#{line_padding(end_line, max_digits)}#{end_line} │ #{line}"
```
Comment (josevalim) [inline lib/elixir/lib/exception.ex:1053]:
```suggestion
    formatted_line = [line_padding(end_line, max_digits), end_line, " | ", line]
```

File: lib/elixir/lib/exception.ex [hunk 12]
```diff
@@ -949,28 +951,174 @@ defmodule MismatchedDelimiterError do
...
+       when start_line < end_line do
+    max_digits = digits(end_line)
+    general_padding = max(2, max_digits) + 1
+    padding = n_spaces(general_padding)
+
+    relevant_lines =
+      if end_line - start_line < @max_lines_shown do
+        line_range(lines, start_pos, end_pos, padding, max_digits)
+      else
+        start_padding = line_padding(start_line, max_digits)
```
Comment (josevalim) [inline lib/elixir/lib/exception.ex:986]: Let's move this branch to a helper function as well , WDYT?

File: lib/elixir/lib/exception.ex [hunk 13]
```diff
@@ -949,28 +951,174 @@ defmodule MismatchedDelimiterError do
...
+        first_line = Enum.fetch!(lines, start_line - 1)
+
+        range = (end_line - 2)..(end_line - 1)
+        last_lines = lines
+                     |> Enum.slice(range)
+                     |> Enum.zip_with(range, fn line, line_number ->
+                      line_number = line_number + 1
+                      line_padding = line_padding(line_number, max_digits)
+                      formatted_line = "#{line_padding}#{line_number} │ #{line}"
```
Comment (josevalim) [inline lib/elixir/lib/exception.ex:995]:
```suggestion
                      formatted_line = [line_padding, line_number, " │ ", line]
```

**Review by josevalim — COMMENTED** [review]
I have dropped some final nitpicks, but overall this looks excellent to me. Ready to merge when you are!

File: lib/elixir/test/elixir/kernel/diagnostics_test.exs (hunk 1 above)
Comment (viniciusmuller) [reply lib/elixir/test/elixir/kernel/diagnostics_test.exs]: Hey I've started implementing the line trimming and I've noticed it's really full edge cases and the implementation would be very tricky if we tried to make it have a consistent behavior for all cases. I took a look to see how rust handles it and they don't :laughing: 
[Here's a playground showcasing how it works ](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f34eeee1c83d0e6dfd54085e178423b4), they simply output the line and don't try to trim it even though it might be a bit long.

After thinking for a while I think that's the best because if we trim it by default users with larger terminals won't be able to see the full lines even though they can, and I think it makes the behavior more consistent with what we have today (currently our `format_snippet` function only trims white spaces in lines and if it's longer than a given threshold.

Let me know your opinions, but I think we can start with simpler code and try to implement this in the future if there's a solid reason to

File: lib/elixir/src/elixir_tokenizer.erl [hunk 14]
```diff
@@ -1419,7 +1419,7 @@ check_terminator({End, {EndLine, EndColumn, _}}, [{Start, {StartLine, StartColum
     ExpectedEnd ->
       Suffix =
         io_lib:format(
-          "\n\n    HINT: the \"~ts\" on line ~B is missing terminator \"~ts\"",
+          "\nHINT: the \"~ts\" on line ~B is missing terminator \"~ts\"",
```
Comment (josevalim) [inline lib/elixir/src/elixir_tokenizer.erl:1422]: We no longer need to show this hint in my opinion and we can get rid of the whole suffix.

File: lib/elixir/test/elixir/kernel/diagnostics_test.exs (hunk 1 above)
Comment (josevalim) [reply lib/elixir/test/elixir/kernel/diagnostics_test.exs]: Yeah, I don't think we need to handle this. At best, we trim whitespace.

**Review by josevalim — COMMENTED** [review]
One comment and don't forget to run `make format`. :)

File: lib/elixir/src/elixir_tokenizer.erl (hunk 14 above)
Comment (viniciusmuller) [reply lib/elixir/src/elixir_tokenizer.erl]: Done, I personally think the hint is indeed a bit redudant, but at the same time we don't show what we were expecting, so maybe for a complete beginner it may not be obvious that a `fn` expects an `end`.

Perhaps in the future we could the missing token from the lexer and show something like `└  mismatched unclosed delimiter (expecting "end")`

Comment (josevalim) [reply lib/elixir/src/elixir_tokenizer.erl]: But I also like your suggestion of bringing that from the lexer and including on the mismatched one. Preferences?

Comment (viniciusmuller) [reply lib/elixir/src/elixir_tokenizer.erl]: I think it looks better on the line that points to the mismatched closing delimiter. The procedure would be to put this info in the location, kinda using it as a Meta and then just pass it to the exception as a field?

Comment (josevalim) [reply lib/elixir/src/elixir_tokenizer.erl]: Yeah, in this case, I think we can add to open_delimiter and close_delimiter, and have both available?

File: lib/elixir/lib/exception.ex [hunk 15]
```diff
@@ -956,14 +959,17 @@ defmodule MismatchedDelimiterError do
         end_line: end_line,
         end_column: end_column,
         description: description,
+        opening_delimiter: _open_delimiter,
+        closing_delimiter: _close_delimiter,
+        expected_closing_delimiter: expected_closing_delimiter,
```
Comment (josevalim) [inline lib/elixir/lib/exception.ex:964]: Nitpick: instead of storing it here, expose `:elixir_tokenizer.terminator` and call it from here. :)

File: lib/elixir/lib/exception.ex [hunk 16]
```diff
@@ -1117,8 +1126,8 @@ defmodule MismatchedDelimiterError do
     |> Enum.intersperse("\n")
   end
 
-  defp mismatched_closing_delimiter(end_column),
-    do: [n_spaces(end_column - 1), red("└ mismatched closing delimiter")]
+  defp mismatched_closing_delimiter(end_column, expected_closing_delimiter),
+    do: [n_spaces(end_column - 1), red(~s/└ mismatched closing delimiter (expecting "#{expected_closing_delimiter}")/)]
```
Comment (josevalim) [inline lib/elixir/lib/exception.ex:1130]:
```suggestion
    do: [n_spaces(end_column - 1), red(~s/└ mismatched closing delimiter (expected "#{expected_closing_delimiter}")/)]
```

Expected reads better no? Since it is "mismatched"?

**Comment by josevalim** [comment]
Ah, we can also add them to undefined local functions :)

**Comment by viniciusmuller** [comment]
> I think the last obvious issue to tackle is to add spans for unused and undefined variables. :) Then we are good to go for now. WDYT?

Nice! I assume there we'll should just need to get the UTF-8 length of the variable names and add a span `{start_col, start_col + var_len}` them to meta, right?
Since we should already be handling the span correctly

**Comment by josevalim** [comment]
Correct and `:string.length` does it correctly. For local function I guess there is a question if we only include the function call or more. If more, then it may actually get quite complex.
